src/resources/rmt_esp32.cc

```diff
@@ -198,42 +272,87 @@ static void flush_buffer(RingbufHandle_t rb) {
+
+  err = rmt_rx_start(channel, true);
+  if (err != ESP_OK) return Primitive::os_error(err, process);
+
+  size_t length = 0;
+  TickType_t timeout_ticks = static_cast<TickType_t>(pdMS_TO_TICKS(timeout_ms));
+  if (timeout_ticks == 0 && timeout_ms != 0) timeout_ticks = 1;
+  void* received_bytes = xRingbufferReceive(rb, &length, timeout_ticks);
+  if (received_bytes != null) {
+    if (length > max_output_len) {
+      length = max_output_len;
+    }
+    ByteArray::Bytes bytes(data);
+    memcpy(bytes.address(), received_bytes, length);
+    vRingbufferReturnItem(rb, received_bytes);
+  }
+
+  err = rmt_rx_stop(channel);
+  if (err != ESP_OK) return Primitive::os_error(err, process);
+  data->resize_external(process, length);
+  return data;
+}
+
 PRIMITIVE(transmit_and_receive) {
-  ARGS(int, tx_num, int, rx_num, Blob, transmit_bytes, Blob, receive_bytes, int, max_output_len, int, receive_timeout)
+  ARGS(RMTResource, tx, RMTResource, rx, Blob, transmit_bytes, Blob, receive_bytes, int, max_output_len, int, receive_timeout_ms)
   if (transmit_bytes.length() % 4 != 0) INVALID_ARGUMENT;
   if (receive_bytes.length() % 4 != 0) INVALID_ARGUMENT;
 
+  rmt_channel_t rx_channel = rx->channel();
+  rmt_channel_t tx_channel = tx->channel();
+
   Error* error = null;
   // Force external, so we can adjust the length after the read.
   ByteArray* data = process->allocate_byte_array(max_output_len, &error, true);
   if (data == null) return error;
 
   const rmt_item32_t* transmit_items = reinterpret_cast<const rmt_item32_t*>(transmit_bytes.address());
   const rmt_item32_t* receive_items = reinterpret_cast<const rmt_item32_t*>(receive_bytes.address());
-  rmt_channel_t rx_channel = (rmt_channel_t) rx_num;
 
   RingbufHandle_t rb = null;
   esp_err_t err = rmt_get_ringbuf_handle(rx_channel, &rb);
   if (err != ESP_OK) return Primitive::os_error(err, process);
 
   flush_buffer(rb);
   if (transmit_bytes.length() > 0) {
-    err = rmt_write_items(static_cast<rmt_channel_t>(tx_num), transmit_items, transmit_bytes.length() / 4, true);
+    err = rmt_write_items(tx_channel, transmit_items, transmit_bytes.length() / 4, true);
     if (err != ESP_OK) return Primitive::os_error(err, process);
   }
   err = rmt_rx_start(rx_channel, true);
   if (err != ESP_OK) return Primitive::os_error(err, process);
 
   if (receive_bytes.length() > 0) {
-    err = rmt_write_items(static_cast<rmt_channel_t>(tx_num), receive_items, receive_bytes.length() / 4, true);
+    err = rmt_write_items(tx_channel, receive_items, receive_bytes.length() / 4, true);
     if (err != ESP_OK) {
       rmt_rx_stop(rx_channel);
       return Primitive::os_error(err, process);
     }
   }
 
+  uint16 thr;
+  rmt_get_rx_idle_thresh(rx_channel, &thr);
+  uint8 cnt;
+  rmt_get_clk_div(rx_channel, &cnt);
+
+
   size_t length = 0;
-  void* received_bytes = xRingbufferReceive(rb, &length, receive_timeout);
+  TickType_t timeout_ticks = static_cast<TickType_t>(pdMS_TO_TICKS(receive_timeout_ms));
```

Comment:
Is it possible to share more of this code with receive?

Reply:
Will do it in a follow-up PR.